plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/N4JSTypeSystem.java

```diff
@@ -321,4 +326,51 @@ public TypeRef createSimplifiedIntersection(List<TypeRef> typeRefs, Resource res
 			return null;
 		}
 	}
+
+	/** @return the return {@link TypeRef} for a given {@link FunctionOrFieldAccessor} */
+	public TypeRef getReturnTypeRef(FunctionOrFieldAccessor fofa) {
+		TypeRef inferredType = tau(fofa);
+		if (inferredType instanceof FunctionTypeExprOrRef) {
+			return getActualReturnTypeRef(fofa, (FunctionTypeExprOrRef) inferredType);
+		}
+		if (inferredType instanceof ParameterizedTypeRef) {
+			return inferredType;
+		}
+		return null;
+	}
+
+	private TypeRef getActualReturnTypeRef(FunctionOrFieldAccessor fofa, FunctionTypeExprOrRef inferredType) {
+		if (inferredType == null) {
+			return null;
+		}
+		TypeRef typeRef = inferredType.getReturnTypeRef();
+		if (fofa instanceof FunctionDefinition) {
+			FunctionDefinition fDef = (FunctionDefinition) fofa;
+			Type tFun = fDef.getDefinedType();
+			if (fofa.isAsync()) {
+				typeRef = getTypeArgumentOfReturnType(fDef, tFun, 0);
+			}
+			if (fDef.isGenerator()) {
+				typeRef = getTypeArgumentOfReturnType(fDef, tFun, 1);
+			}
+		}
+		return typeRef;
+	}
+
+	private TypeRef getTypeArgumentOfReturnType(FunctionOrFieldAccessor functionOrFieldAccessor, Type type, int idx) {
+		if (type instanceof TFunction) {
+			TFunction tFun = (TFunction) type;
+			TypeRef actualReturnTypeRef = tFun.getReturnTypeRef();
+			if (actualReturnTypeRef.getTypeArgs().size() > idx) {
+				TypeArgument tReturn = actualReturnTypeRef.getTypeArgs().get(idx);
+				RuleEnvironment fofaRuleEnv = RuleEnvironmentExtensions.newRuleEnvironment(functionOrFieldAccessor);
+				TypeRef typeRef = resolveType(fofaRuleEnv, tReturn);
+				if (TypeUtils.isUndefined(typeRef)) {
+					typeRef = RuleEnvironmentExtensions.voidTypeRef(fofaRuleEnv);
```

Comment:
for this we already have `TypeSystemHelper#sanitizeTypeOfVariableFieldProperty(RuleEnvironment, TypeArgument)` (which also handles some other special cases)